test/dhtest.c

```diff
@@ -62,11 +62,15 @@ static int dh_test(void)
         || !TEST_true(DH_set0_pqg(dh, p, q, g)))
         goto err;
 
+    p1=p, p=NULL;
```

Comment:
I think I'm horribly confused -- isn't ```p``` still a ```BN_new()```-allocated object here?  I don't see ```p1``` getting freed, so it looks like it's leaking...

Reply:
If I understand `DH_set0_pqg()` correctly, the three variables p, q, and g are assigned to members of the `dh` object without incrementing the reference count, so they are now owned by the dh object.

Reply:
Ah, yes, that would be the piece I was missing -- thanks!
("set0" as in "0 additional references/copies are made")

Reply:
Yes, it's like @mspncp says, but as @kaduk pointed, its' a bit confusing!

I suggest (https://github.com/openssl/openssl/pull/7155#discussion_r216167089) to add a comment to clarify (unless we go with the alternative solution https://github.com/openssl/openssl/pull/7155#discussion_r216167448)

Reply:
@romen I‘m tending towards your alternative proposal.